### ISSUE TITLE
Permissions issues when building with Docker

### ISSUE THREAD

**Issue by nopid** [iodide-project/pyodide#865]
The user is root Inside the container launnched by `run_docker` script. This introduces subtle issues and build hiccups.
https://github.com/iodide-project/pyodide/blob/3a06f5dfcb9b536e9ece1f68f6963717acf82486/run_docker#L72-L78

When run as root, the `tar` utility used in several packages Makefile will, by default, preserves id and permissions when extracting the archives. It might generate permissions errors.

On MacOS, from master (3a06f5dfcb9b536e9ece1f68f6963717acf82486), a simple build fails as such:
```
$ ./run_docker --pre-built
root@f8a91d8b89ac:/src#  make clean-all
(...)
root@f8a91d8b89ac:/src# PYODIDE_PACKAGES="micropip,pillow" make
(...)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/shutil.py", line 788, in move
    os.rename(src, real_dst)
OSError: [Errno 18] Invalid cross-device link: '/tmp/emscripten_temp_crlea8jq/xlink_0.o' -> '/src/packages/libxml/libxml2-2.9.10/xlink.o'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/src/emsdk/emsdk/fastcomp/emscripten/emcc.py", line 3548, in <module>
    sys.exit(run(sys.argv))
  File "/src/emsdk/emsdk/fastcomp/emscripten/emcc.py", line 1916, in run
    safe_move(temp_file, bitcode_target)
  File "/src/emsdk/emsdk/fastcomp/emscripten/tools/shared.py", line 3183, in safe_move
    shutil.move(src, dst)
  File "/usr/local/lib/python3.8/shutil.py", line 802, in move
    copy_function(src, real_dst)
  File "/usr/local/lib/python3.8/shutil.py", line 433, in copy2
    copystat(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/local/lib/python3.8/shutil.py", line 372, in copystat
    lookup("utime")(dst, ns=(st.st_atime_ns, st.st_mtime_ns),
OSError: [Errno 5] Input/output error
make[4]: *** [Makefile:1291: xlink.lo] Error 1
make[4]: *** Waiting for unfinished jobs....
make[4]: Leaving directory '/src/packages/libxml/libxml2-2.9.10'
make[3]: *** [Makefile:1479: all-recursive] Error 1
make[3]: Leaving directory '/src/packages/libxml/libxml2-2.9.10'
make[2]: *** [Makefile:893: all] Error 2
make[2]: Leaving directory '/src/packages/libxml/libxml2-2.9.10'
make[1]: *** [Makefile:35: /src/packages/libxml/libxml2-2.9.10/.libs/libxml2.a] Error 2
make[1]: Leaving directory '/src/packages/libxml'
make: *** [Makefile:248: packages/libxml/libxml2-2.9.10/.libs/libxml2.a] Error 2
```

Looking at permissions:
```
root@f8a91d8b89ac:/src# ls -l packages/libxml/
total 8
-rw-r--r--   1 root root  994 Dec 14 09:44 Makefile
-rw-r--r--   1 root root   94 Dec 14 09:44 checksums
drwxr-xr-x   3 root root   96 Dec 15 13:50 downloads
drwxrwxr-x 192 1000 1000 6144 Dec 15 13:56 libxml2-2.9.10
```

The `libxml2-2.9.10` directory is owned by 1000:1000 without others writing permissions, hence previous OSError.

A simple fix for this problem would be to replace `tar` by `tar -o` in every Makefile.

A long time fix would be to run as a sudoer user inside docker to avoid root issues.

**Comment by rth** [comment]
Thanks for the report. The current situation with permissions when using Docker is indeed not ideal.

BTW, don't use pre-built Docker images for development, they are a few weeks outdated and use an older version of emsdk. I will regenerate them for the release, but meanwhile using the non pre-built version is preferable.

**Comment by rth** [comment]
For instance to build pillow in the non pre-built version, one would do,
```
PYODIDE_PACKAGES="pillow" make
```
That shouldn't take too long.

For scikit-image, one has currently to manually build CLAPACK,
```
make -C packages/CLAPACK clean
make -C packages/CLAPACK
```
when using `PYODIDE_PACKAGES` for packages that depend on LAPACK (i.e. scipy).

**Comment by nopid** [comment]
It turns out I have plenty of OSError 5 input/output error when trying to build pyodide. It might be a Docker bug. I will close this issue.

**Comment by rth** [comment]
As to the `Invalid cross-device link` errors it looks like emcc is trying to create links between files in /tmp/ (filesystem inside Docker) and /src/packages (locally mounted file-system). It's very likely that this is specific to Docker on MacOS, though I'm not sure if other people managed to make it work on that platform.

The other solution could be try building without Docker. I would be very curious if something like https://github.com/iodide-project/pyodide/pull/830#pullrequestreview-546619206 would work on MacOS. It might be necessary to apply https://github.com/iodide-project/pyodide/pull/590 first though.

**Comment by nopid** [comment]
The `Invalid cross-device link` error is catched. Indeed it is `shutil.move` that is used (way better than `os.rename`). It first tries `os.rename` and in case of error it falls back to `shutil.move`. I first suspected this problem to be the cause of my multiple errors but there are only 4 places inside the source of pyodide where `os.rename` is still in use (this could be fixed and replaced by `shutil.move`).

I pruned my docker system and installed the last version. If it doesn't build I will try to find another option.
